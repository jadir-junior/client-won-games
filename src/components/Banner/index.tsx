import * as S from './styles'

import Ribbon, { RibbonColors, RibbonSizes } from 'components/Ribbon'

import Button from 'components/Button'
import Image from 'next/image'

export type BannerProps = {
  img: string
  title: string
  subtitle: string
  buttonLabel: string
  buttonLink: string
  ribbon?: React.ReactNode
  ribbonSize?: RibbonSizes
  ribbonColor?: RibbonColors
}

const Banner = ({
  img,
  title,
  subtitle,
  buttonLabel,
  buttonLink,
  ribbon,
  ribbonSize,
  ribbonColor
}: BannerProps) => (
  <S.Wrapper>
    {!!ribbon && (
      <Ribbon size={ribbonSize} color={ribbonColor}>
        {ribbon}
      </Ribbon>
    )}
    <S.ImageWrapper>
      <Image src={img} alt={title} layout="fill" objectFit="cover" />
    </S.ImageWrapper>
    <S.Caption>
      <S.Title>{title}</S.Title>
      <S.SubTitle dangerouslySetInnerHTML={{ __html: subtitle }} />
      <Button as="a" href={buttonLink}>
        {buttonLabel}
      </Button>
    </S.Caption>
  </S.Wrapper>
)

export default Banner
