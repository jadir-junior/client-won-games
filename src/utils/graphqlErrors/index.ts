import { ApolloError } from '@apollo/client'

export function graphqlErrors(error: ApolloError) {
  return error?.graphQLErrors[0]?.extensions?.exception.data.message[0]
    .messages[0].message
}
