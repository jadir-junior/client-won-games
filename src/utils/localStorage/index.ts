const APP_KEY = 'WONGAMES'

export function getStorageItem(key: string) {
  // no Next via Server/Static não tem window
  if (typeof window === 'undefined') return

  return JSON.parse(window.localStorage.getItem(`${APP_KEY}_${key}`)!)
}

export function setStorageItem(key: string, data: string[]) {
  // no Next via Server/Static não tem window
  if (typeof window === 'undefined') return

  return window.localStorage.setItem(`${APP_KEY}_${key}`, JSON.stringify(data))
}
