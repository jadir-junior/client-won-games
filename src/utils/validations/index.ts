import Joi from 'joi'
import { UsersPermissionsRegisterInput } from 'graphql/generated/globalTypes'

const fieldsValidations = {
  username: Joi.string().min(5).required(),
  email: Joi.string().email({ tlds: false }).required(),
  password: Joi.string().required(),
  confirm_password: Joi.string()
    .valid(Joi.ref('password'))
    .required()
    .messages({ 'any.only': 'confirm password does not match with password' })
}

export type FieldErrors = {
  [key: string]: string
}

function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

export function signUpValidate(values: UsersPermissionsRegisterInput) {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type SingInValues = Omit<UsersPermissionsRegisterInput, 'username'>
export function signInValidate(values: SingInValues) {
  const { email, password } = fieldsValidations

  const schema = Joi.object({ email, password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type ForgetValidadeParams = Pick<UsersPermissionsRegisterInput, 'email'>
export function forgotValidate(values: ForgetValidadeParams) {
  const { email } = fieldsValidations

  const schema = Joi.object({ email })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type ResetValidadeParams = {
  password: string
  confirm_password: string
}
export function resetValidate(values: ResetValidadeParams) {
  const { password, confirm_password } = fieldsValidations

  const schema = Joi.object({ password, confirm_password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
