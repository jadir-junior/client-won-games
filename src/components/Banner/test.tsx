import { render, screen } from 'utils/test-utils'

import Banner from '.'

const props = {
  img: 'https://source.unsplash.com/user/willianjusten/1042x580',
  title: 'Defy death',
  subtitle: '<p>Play the new <strong>CrashLands</strong> season',
  buttonLabel: 'Buy now',
  buttonLink: '/games/defy-death'
}

describe('<Banner />', () => {
  it('should render the Banner', () => {
    const { container } = render(<Banner {...props} />)

    expect(
      screen.getByRole('heading', { name: /defy death/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /play the new crashlands season/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /defy death/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a Ribbon', () => {
    render(
      <Banner
        {...props}
        ribbon="My Ribbon"
        ribbonSize="small"
        ribbonColor="secondary"
      />
    )

    const ribbon = screen.getByText(/My Ribbon/i)

    expect(ribbon).toBeInTheDocument()
    expect(ribbon).toHaveStyle({ backgroundColor: '#3CD3C1' })
    expect(ribbon).toHaveStyle({
      height: '2.6rem',
      fontSize: '1.2rem'
    })
  })
})
