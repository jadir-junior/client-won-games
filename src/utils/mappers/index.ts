import {
  QueryHome_banners,
  QueryHome_newGames,
  QueryHome_sections_popularGames_highlight
} from 'graphql/generated/QueryHome'

import { QueryGames_games } from 'graphql/generated/QueryGames'
import { QueryOrders_orders } from 'graphql/generated/QueryOrders'
import formatPrice from 'utils/format-price'
import { getImageUrl } from 'utils/getImageUrl'

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map((banner) => ({
    img: `${getImageUrl(banner.image?.url)}`,
    title: banner.title,
    subtitle: banner.subtitle,
    buttonLabel: banner.button?.label,
    buttonLink: banner.button?.link,
    ...(banner.ribbon && {
      ribbon: banner.ribbon.text,
      ribbonColor: banner.ribbon.color,
      ribbonSize: banner.ribbon.size
    })
  }))
}

export const gamesMapper = (games: QueryHome_newGames[] | null | undefined) => {
  return games
    ? games.map((game) => ({
        id: game.id,
        title: game.name,
        slug: game.slug,
        img: `${getImageUrl(game.cover?.url)}`,
        developer: game.developers[0].name,
        price: game.price
      }))
    : []
}

export const highlightMapper = (
  highlight: QueryHome_sections_popularGames_highlight | null | undefined
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `${getImageUrl(highlight.background?.url)}`,
        floatImage: `${getImageUrl(highlight.floatImage?.url)}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment
      }
    : {}
}

export const cartMapper = (items: QueryGames_games[] | null | undefined) => {
  return items
    ? items.map((item) => ({
        id: item.id,
        img: `${getImageUrl(item.cover?.url)}`,
        price: formatPrice(item.price),
        title: item.name
      }))
    : []
}

export const ordersMapper = (
  orders: QueryOrders_orders[] | null | undefined
) => {
  return orders
    ? orders.map((order) => {
        return {
          id: order.id,
          paymentInfo: {
            flag: order.card_brand ? order.card_brand : null,
            img: order.card_brand ? `/img/cards/${order.card_brand}.png` : null,
            number: order.card_last4
              ? `**** **** **** ${order.card_last4}`
              : 'Free Game',
            purchaseDate: `Purchase made on ${new Intl.DateTimeFormat('en-US', {
              day: 'numeric',
              month: 'short',
              year: 'numeric'
            }).format(new Date(order.created_at))}`
          },
          games: order.games.map((game) => ({
            id: game.id,
            title: game.name,
            downloadLink:
              'https://wongames.com/game/download/yuYT56Tgh431LkjhNBgdf',
            img: `${getImageUrl(game.cover?.url)}`,
            price: formatPrice(game.price)
          }))
        }
      })
    : []
}
