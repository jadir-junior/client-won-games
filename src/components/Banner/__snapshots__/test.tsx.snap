// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render the Banner 1`] = `
.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  cursor: pointer;
  outline: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c4:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  background-image: url(https://source.unsplash.com/user/willianjusten/1042x580);
  background-position: center center;
  background-size: cover;
}

.c1 {
  width: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 2.4rem;
}

.c2 {
  font-size: 1.8rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c3 {
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 400;
  margin-bottom: 1.6rem;
}

.c3 strong {
  color: #F231A5;
  font-weight: 600;
}

@media (min-width:768px) {
  .c0 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c1 {
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
    position: absolute;
    bottom: 0;
    left: 0;
  }
}

@media (min-width:768px) {
  .c2 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 1.8rem;
  }
}

<div
  class=""
>
  <div
    aria-label="Defy death"
    class="c0"
    role="img"
    src="https://source.unsplash.com/user/willianjusten/1042x580"
  />
  <div
    class="c1"
  >
    <h2
      class="c2"
    >
      Defy death
    </h2>
    <h3
      class="c3"
    >
      <p>
        Play the new 
        <strong>
          CrashLands
        </strong>
         season
      </p>
    </h3>
    <a
      class="c4"
      href="/games/defy-death"
    >
      <span>
        Buy now
      </span>
    </a>
  </div>
</div>
`
